src/dune_engine/build_config.mli

```diff
@@ -14,13 +14,27 @@ end
 
 module Rules : sig
   (** Rules for a given directory. This type is structured so that all generated
-      sub-directories (either directory targets or internal generated
+      sub-directories (either directory targets or internally generated
       directories such as [.ppx]) are known immediately, while the actual build
       rules are computed in a second stage. The staging is to avoid computation
       cycles created during the computation of the rules. *)
 
+  module Build_only_sub_dirs : sig
+    (** The set of either directory targets or internally generated directories,
+        indexed by their parent build directory. *)
+    type t
+
+    val empty : t
+
+    val singleton : dir:Path.Build.t -> Subdir_set.t -> t
+
+    val find : t -> Path.Build.t -> Subdir_set.t option
```

Comment:
I feel like there isn't any meaningful distinction between `None` and `Some Subdir_set.empty` outcomes, which suggests that we can get rid of the `option` here.

Reply:
[c000eb5](https://github.com/ocaml/dune/pull/7207/commits/c000eb5cd7eb19c88193d7f6c387edd55e7cad4b)